packages/components/src/color-picker/inputs.js

```diff
@@ -0,0 +1,287 @@
+		}
+		this.setState( { value } );
+	}
+
+	handleKeyDown( { keyCode } ) {
+		if ( keyCode !== ENTER && keyCode !== UP && keyCode !== DOWN ) {
+			return;
+		}
+		const { value } = this.state;
+		const { label, onChange } = this.props;
+		const propName = String( label ).toLowerCase();
+		onChange( { [ propName ]: value } );
+	}
+
+	render() {
+		const { label, ...props } = this.props;
+		const { value } = this.state;
+		return (
+			<TextControl
+				className="components-color-picker__inputs-field"
+				label={ label }
+				value={ value }
+				onChange={ ( newValue ) => this.handleChange( newValue ) }
+				onBlur={ this.handleBlur }
+				onKeyDown={ this.handleKeyDown }
+				{ ...omit( props, [ 'onChange', 'value' ] ) }
+			/>
+		);
+	}
+}
+
+export class Inputs extends Component {
+	constructor( { hsl } ) {
+		super( ...arguments );
+
+		const view = hsl.a === 1 ? 'hex' : 'rgb';
+		this.state = { view };
+
+		this.toggleViews = this.toggleViews.bind( this );
+		this.handleChange = this.handleChange.bind( this );
+	}
+
+	static getDerivedStateFromProps( props, state ) {
+		if ( props.hsl.a !== 1 && state.view === 'hex' ) {
+			return { view: 'rgb' };
+		}
+		return null;
+	}
+
+	toggleViews() {
+		if ( this.state.view === 'hex' ) {
+			this.setState( { view: 'rgb' } );
+		} else if ( this.state.view === 'rgb' ) {
+			this.setState( { view: 'hsl' } );
+		} else if ( this.state.view === 'hsl' ) {
+			if ( this.props.hsl.a === 1 ) {
+				this.setState( { view: 'hex' } );
+			} else {
+				this.setState( { view: 'rgb' } );
+			}
+		}
+	}
+
+	handleChange( data ) {
+		if ( data.hex ) {
+			if ( isValidHex( data.hex ) ) {
+				this.props.onChange( {
+					hex: data.hex,
+					source: 'hex',
+				} );
+			}
+		} else if ( data.r || data.g || data.b ) {
+			this.props.onChange( {
+				r: data.r || this.props.rgb.r,
+				g: data.g || this.props.rgb.g,
+				b: data.b || this.props.rgb.b,
+				source: 'rgb',
+			} );
+		} else if ( data.a ) {
+			if ( data.a < 0 ) {
+				data.a = 0;
+			} else if ( data.a > 1 ) {
+				data.a = 1;
+			}
+
+			this.props.onChange( {
+				h: this.props.hsl.h,
+				s: this.props.hsl.s,
+				l: this.props.hsl.l,
+				a: Math.round( data.a * 100 ) / 100,
+				source: 'rgb',
+			} );
+		} else if ( data.h || data.s || data.l ) {
+			// Remove any occurances of '%'.
+			if ( typeof data.s === 'string' && data.s.includes( '%' ) ) {
+				data.s = data.s.replace( '%', '' );
+			}
+			if ( typeof data.l === 'string' && data.l.includes( '%' ) ) {
+				data.l = data.l.replace( '%', '' );
+			}
+
+			this.props.onChange( {
+				h: data.h || this.props.hsl.h,
+				s: Number( ( data.s && data.s ) || this.props.hsl.s ),
+				l: Number( ( data.l && data.l ) || this.props.hsl.l ),
+				source: 'hsl',
+			} );
+		}
+	}
+
+	renderFields() {
+		const { disableAlpha = false } = this.props;
+		if ( this.state.view === 'hex' ) {
+			return (
+				<fieldset>
+					<legend className="screen-reader-text">
+						{ __( 'Color value in HEX' ) }
```

Comment:
Minor: Is "HEX" a word that ought be capitalized?

Reply:
I'd say no. It is not an acronym like RGB ("red-green-blue"); it is an abbreviation of "hexadecimal".

Actually, why not just say "Color value in hexadecimal"?

Reply:
"Color value in hexadecimal" works for me.

Reply:
When there's only one input field, a fieldset + legend are not so appropriate. They're meant to group a set of controls and give the group a name. It is redundant and potentially confusing for users, as they would hear there's a group when actually there's no group:

<img width="636" alt="screen shot 2018-10-17 at 16 13 22" src="https://user-images.githubusercontent.com/1682452/47093730-13f3bc00-d22a-11e8-9d85-3b3347d3e333.png">

I'd suggest to remove fieldset + legend for the hex color and use "Color value in hexadecimal" as the visible `<label>` element of the field.

Reply:
Fixed at 8490b97